Support for quoted types, pythonic script generation, strict DATE regex

jtschichold opened the pull request:
## Motivation

This PR includes 3 different changes I found useful:
- If you load on BigQuery a JSON objects where INTEGERs, FLOATs and BOOLEANs are _stringified_ (`"1"` instead of `1`), BigQuery will detect the real type of the fields and mark them as INTEGER, FLOAT and BOOLEAN in the autodetected schema. bigquery-schema-generation doesn't support this type of autodetection and marks all the stringified fields as STRINGs. This PR enhances the infer mechanism to detect _stringified_ fields.
- BigQuery performs additional checks on the DATEs field not performed by bigquery-schema-generation. This means that some times load operation files because some of the dates are invalid example: `1988-00-00`. This PR improves the regex for DATEs detection.
- The current generate-schema script uses a python3 without any PATH. This is incompatibile with venv. This PR change the script install mechanism to use the _console_scripts_ entrypoint feature of setuptools (https://python-packaging.readthedocs.io/en/latest/command-line-scripts.html)

## Modifications

- Added 3 new string types called `QINTEGER`, `QFLOAT`, `QBOOLEAN` to represent integers, float and booleans detected inside STRINGs.
- Changed the DATE regex to make it more strict
- Changed setup.py and added a console_scripts keyword in the setup call to automatically generate the generate-schema script. Removed the old script.

## Result

- _stringified_ fields are now supported
- only proper DATEs are used as DATEs :-)
- generate-schema script is automatically generated by setuptools and references the right python binary

## Note
I know it's not good form including 3 different changes in the same PR. Let me know if you want me to split them over 3 different PRs.

bxparks commented:
Hi Luigi,

Thanks for putting this together, these are great improvements. The PR is small enough that I think we can do this in one PR.

Do you have time to update the unit tests in `tests/test_generate_schema.py`? Just the basic ones (test_date_matcher_valid(), test_date_matcher_invalid(), test_infer_value_type(), test_infer_bigquery_type(), test_infer_array_type(), test_convert_type()). I can handle the far more complicated `TestFromDataFile` class in a separate pass.

I have a quick design question though: Do you actually need to create the QINTEGER, QFLOAT, and QBOOLEAN types? Once you detect a quoted numerical value, you might be able to map it directly to the existing INTEGER, FLOAT, and BOOLEAN types, and the rest of the logic will just happen. If so, this might eliminate 10-12 lines of your code. The unit tests will verify if my theory holds.

Brian

jtschichold commented:
Hi Brian,
sorry, I missed the unit tests. I have added a new commit to:
- fix an issue with the date regex detected by your unit tests
- added a specific unit test for the incorrect date `1988-00-00`
- added to testdata.txt some tests on Q* types. These tests should show why I didn't use the INTEGER, BOOLEAN and FLOAT types. Check this:
```
# QBOOLEAN -> STRING
DATA
{ "qb" : "true" }
{ "qb" : "False" }
{ "qb" : "test" }
SCHEMA
[
  {
    "mode": "NULLABLE",
    "name": "qb",
    "type": "STRING"
  }
]
END
```
For my understanding, if I used BOOLEAN instead of a new QBOOLEAN the existing logic wouldn't have converted the field from BOOLEAN to STRING

If you think Q* types are necessary I will go ahead and all the unit tests to cover the regexs.

Thanks,
Luigi

bxparks commented:
Ah yes, I see what you mean. A QINTEGER + STRING must convert to a STRING. It seems then that the `convert_type()` method needs to be far more complex, because it needs to also handle mixed conversions like (QINTEGER + FLOAT => FLOAT), and (QFLOAT + INTEGER => FLOAT), and various permutations of those. And also, it needs to handle mixed conversions with the date/time types: e.g. (QINTEGER + TIMESTAMP => STRING).

I verified that `bq load` supports these mixed conversions.

I created the `bxparks-support-for-quoted-types` branch on my repository with a commit that I think contains all these mixed conversions. Also added extensive unit tests that should cover all permutations. Can you take a look? If that looks good, I think you should be able to merge my branch into your branch, then push your branch again to update your PR. I can merge your PR into `develop`.

P.S. I'm going to be travelling for the next 4-5 days, so will be slow to respond.

jtschichold commented:
Hi Brian,
awesome! I will work on an additional commit with unit tests for the Q* REs

bxparks commented:
Sorry for delay. Merged. Things look pretty good, but I won't know for certain until I try to upload to PyPI.

Sounds like you have some other changes that you'd like merged?
